config/src/main/java/com/typesafe/config/impl/ConfigImpl.java

```diff
@@ -364,15 +364,37 @@ public static Config defaultReference(final ClassLoader loader) {
         return computeCachedConfig(loader, "defaultReference", new Callable<Config>() {
             @Override
             public Config call() {
-                Config unresolvedResources = Parseable
-                        .newResources("reference.conf",
-                                ConfigParseOptions.defaults().setClassLoader(loader))
-                        .parse().toConfig();
+                Config unresolvedResources = unresolvedReference(loader);
                 return systemPropertiesAsConfig().withFallback(unresolvedResources).resolve();
             }
         });
     }
 
+    private static Config unresolvedReference(final ClassLoader loader) {
+        return computeCachedConfig(loader, "unresolvedReference", new Callable<Config>() {
+            @Override
+            public Config call() {
+                return Parseable.newResources("reference.conf",
+                        ConfigParseOptions.defaults().setClassLoader(loader))
+                    .parse().toConfig();
+            }
+        });
+    }
+
+    /**
+     * This returns the unresolved reference configuration, but before doing so,
+     * it verifies that the reference configuration resolves, to ensure that it
+     * is self contained and doesn't depend on any higher level configuration
+     * files.
+     */
+    public static Config verifiedUnresolvedReference(final ClassLoader loader) {
+        // First, verify that `reference.conf` resolves by itself.
+        defaultReference(loader);
```

Comment:
The exception this throws may be a bit confusing; if you've defined `foo` in application.conf but not `reference.conf` it's going to say "couldn't resolve foo" or whatever with no explanation of the convoluted thing that yeah you can override it in application.conf but the reference config needs to be standalone-resolveable also.

I wonder if it would be helpful to catch the exception and add some detail like "reference.conf must have a default value for all substitutions" - could still use the same `UnresolvedSubstitution` exception type, but with an extra line of text perhaps, not sure how annoying it is to do.

Reply:
I've done this, I had to hack `UnresolvedSubstitution` a little bit to create a sensible error message, since it was losing the detail in a message string, and always prepending its own description to the passed in detail.